zap/src/output/luau/mod.rs

```diff
@@ -18,69 +15,526 @@ pub trait Output {
+
+						self.push_line(&format!("alloc({})", numty.size()));
+						self.push_line(&format!("buffer.write{numty}(outgoing_buff, outgoing_apos, {i})"));
+
+						for (field_name, field_ty) in &variant_struct.fields {
+							self.push_ser(&format!("{from}.{field_name}"), field_ty, checks);
+						}
+					}
+
+					self.push_dedent_line_indent("else");
+					self.push_line("error(\"invalid variant value\")");
+					self.push_dedent_line("end");
+				}
 			},
 
-			Stmt::NumFor { var, from, to } => self.push_line(&format!("for {var} = {from}, {to} do")),
-			Stmt::GenFor { key, val, obj } => self.push_line(&format!("for {key}, {val} in {obj} do")),
-			Stmt::If(cond) => self.push_line(&format!("if {cond} then")),
-			Stmt::ElseIf(cond) => self.push_line(&format!("elseif {cond} then")),
-			Stmt::Else => self.push_line("else"),
-
-			Stmt::End => self.push_line("end"),
-		};
-
-		if matches!(
-			stmt,
-			Stmt::NumFor { .. } | Stmt::GenFor { .. } | Stmt::If(..) | Stmt::ElseIf(..) | Stmt::Else
-		) {
-			self.indent();
-		};
+			Ty::Struct(struct_ty) => {
+				for (field_name, field_ty) in &struct_ty.fields {
+					self.push_ser(&format!("{from}.{field_name}"), field_ty, checks);
+				}
+			}
+
+			Ty::Instance(class) => {
+				if checks && class.is_some() {
+					self.push_line(&format!("assert({from}:IsA(\"{}\"))", class.unwrap()));
+				}
+
+				self.push_line(&format!("table.insert(outgoing_inst, {from})"));
+			}
+
+			Ty::Color3 => {
+				self.push_line("alloc(3)");
+				self.push_line(&format!("buffer.writeu8(outgoing_buff, outgoing_apos, {from}.R * 255)"));
+				self.push_line(&format!(
+					"buffer.writeu8(outgoing_buff, outgoing_apos + 1, {from}.G * 255)"
+				));
+				self.push_line(&format!(
+					"buffer.writeu8(outgoing_buff, outgoing_apos + 1, {from}.B * 255)"
+				));
+			}
+
+			Ty::Vector3 => {
+				self.push_line("alloc(12)");
+				self.push_line(&format!("buffer.writef32(outgoing_buff, outgoing_apos, {from}.X)"));
+				self.push_line(&format!("buffer.writef32(outgoing_buff, outgoing_apos + 4, {from}.Y)"));
+				self.push_line(&format!("buffer.writef32(outgoing_buff, outgoing_apos + 8, {from}.Z)"));
+			}
+
+			Ty::AlignedCFrame => {
+				self.push_line(&format!(
+					"local axis_alignment = table.find(CFrameSpecialCases, {from}.Rotation)"
+				));
+				self.push_line("assert(axis_alignment)");
+				self.push_line("alloc(1)");
+				self.push_line("buffer.writeu8(outgoing_buff, outgoing_apos, axis_alignment)");
+
+				self.push_line("alloc(12)");
+				self.push_line(&format!(
+					"buffer.writef32(outgoing_buff, outgoing_apos, {from}.Position.X)"
+				));
+				self.push_line(&format!(
+					"buffer.writef32(outgoing_buff, outgoing_apos + 4, {from}.Position.Y)"
+				));
+				self.push_line(&format!(
+					"buffer.writef32(outgoing_buff, outgoing_apos + 8, {from}.Position.Z)"
+				));
+			}
+
+			Ty::CFrame => {
+				self.push_line(&format!("local axis, angle = {from}:ToAxisAngle()"));
+				self.push_line("axis = axis * angle");
```

Comment:
I'm not sure how I feel about reassigning here

Reply:
it uses 16 bytes less memory, and it is what it generated before iirc

Reply:
Ok nw